proposals/csharp-7.0/tuples.md

```diff
@@ -0,0 +1,437 @@
+
+```csharp
+(string name, byte age) t = (null, 5); // Ok: the expressions null and 5 convert to string and byte
+```
+
+In cases where the tuple literal is not part of a conversion, it acquires its "natural type", which means a tuple type where the element types are the types of the constituent expressions. Since not all expressions have types, not all tuple literals have a natural type either:
+
+```csharp
+var t = ("John", 5); // Ok: the type of t is (string, int)
+var t = (null, 5); //   Error: null doesn't have a type
+```
+
+A tuple literal may include names, in which case they become part of the natural type:
+
+```csharp
+var t = (name: "John", age: 5); // The type of t is (string name, int age)
+```
+
+Tuple literals may be deconstructed directly:
+
+```csharp
+(string x, byte y, var z) = (null, 1, 2);
+(string x, byte y) t = (null, 1);
+```
+
+Or for deconstructing assignment:
+
+```csharp
+string x;
+byte y;
+
+(x, y) = (null, 1);
+(x, y) = (y, x); // swap!
+```
+
+The evaluation order of deconstruction assignment expressions is "breadth first":
+
+1. Evaluate the LHS: Evaluate each of the expressions inside of it one by one, left to right, to yield side effects and establish a storage location for each.
+1. Evaluate the RHS: Evaluate each of the expressions inside of it one by one, left to right to yield side effects
+1. Convert each of the RHS expressions to the LHS types expected, one by one, left to right.
+1. Assign each of the conversion results from 3 to the storage locations found in 
+
+> **Note to reviewers**: I found this in the LDM notes for July 13-16, 2016. I don't think it is still accurate:
+
+A deconstructing assignment is a *statement-expression* whose type could be `void`.
+
+### Duality with underlying type
+
+Tuples map to underlying types of particular names.
+
+```csharp
+System.ValueTuple<T1, T2>
+System.ValueTuple<T1, T2, T3>
+...
+System.ValueTuple<T1, T2, T3,..., T7, TRest>
+```
+
+Tuple types behave exactly like underlying types. The only additional enhancement is the more expressive field names given by the programmer.
+
+```csharp
+var t = (sum: 0, count: 1);
+t.sum   = 1;        // sum   is the name for the field #1
+t.Item1 = 1;        // Item1 is the name of underlying field #1 and is also available
+
+var t1 = (0, 1); // tuple omits the field names.
+t.Item1 = 1;     // underlying field name is still available
+
+t.ToString();     // ToString on the underlying tuple type is called.
+
+System.ValueTuple<int, int> vt = t; // identity conversion
+(int moo, int boo) t2 = vt;         // identity conversion
+
+```
+
+Because of the dual nature of tuples, it is not allowed to assign field names that overlap with preexisting member names of the underlying type. The only exception is the use of predefined `Item1`, `Item2`,...`ItemN` at corresponding position N, since that would not be ambiguous.
```

Comment:
>with preexisting member names of the underlying type [](start = 91, length = 52)

Not quite true.  We permit, for example, ObjectEquals and GetType.  [https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQMwAJboMLoN7LpGYZQAs6AsgBQCU+hxTA+ugLzrUDyARgFYBTAMYAXAKIBHAK4BDADYBnEOjgAadAHEBIgCoBPAA4DlMWgG5GRAL7IrQA=](https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQMwAJboMLoN7LpGYZQAs6AsgBQCU+hxTA+ugLzrUDyARgFYBTAMYAXAKIBHAK4BDADYBnEOjgAadAHEBIgCoBPAA4DlMWgG5GRAL7IrQA=)

Reply:
Rex has updated this description. I've noted (in our working draft) that the language used may still need refinement. We need to define the term "element" to mean "ItemN" and "Rest" elements, excluding other members.